File: ast/statement/for.go
package statement

import (
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/types"
)

type For struct {
	StatementBase
	Initialization core.Statement
	Condition      core.Expression
	Post           core.Statement
	Body           core.Statement
}

func (f *For) Validate(c core.Context) {
	ctx := c.NewContext()
	if f.Initialization != nil {
		f.Initialization.Validate(ctx)
	}
	if f.Condition != nil {
		conditionCtx := ctx.NewContext()
		f.Condition.Validate(conditionCtx, types.TypeBool)
		if f.Condition.Type() != nil && !f.Condition.Type().Equal(types.TypeBool) {
			c.Error(f.Condition.GetPosition(), "expect bool type condition")
		}
	}
	if f.Post != nil {
		postCtx := ctx.NewContext()
		f.Post.Validate(postCtx)
	}
	if f.Body != nil {
		bodyCtx := ctx.NewContext()
		f.Body.Validate(bodyCtx)
	}
}

File: MicroPanda/AST/Statement/For.cs
namespace MicroPanda.AST.Statement;

using Expression;
using Type;

internal class For : Statement
{
    internal Statement? Initialization { get; set; }
    internal Expression? Condition { get; set; }
    internal Statement? Post { get; set; }
    internal Statement? Body { get; set; }

    internal override void Validate(Context context)
    {
        var ctx = context.NewContext();
        Initialization?.Validate(ctx);
        if (Condition != null)
        {
            var conditionCtx = ctx.NewContext();
            Condition.Validate(conditionCtx, TypeHelper.TypeBool);
            if (Condition.Type != null && !Condition.Type.Equal(TypeHelper.TypeBool))
            {
                context.Program.Error(Condition.Position, "expect bool type condition");
            }
        }
        if (Post != null)
        {
            var postCtx = ctx.NewContext();
            Post.Validate(postCtx);
        }
        if (Body != null)
        {
            var bodyCtx = ctx.NewContext();
            Body.Validate(bodyCtx);
        }
    }
}